examples/rl-ppo-continuous.rs

```diff
@@ -0,0 +1,241 @@
+//! Implements the reinforcement learning algorithm Proximal Policy Optimization (PPO) on random data.
+
+use dfdx::{
+    gradients::Tape,
+    nn::{
+        self,
+        modules::{Linear, ReLU, Tanh},
+        BuildModule, Module,
+    },
+    optim::{Adam, AdamConfig},
+    prelude::{smooth_l1_loss, GradientUpdate, ParamUpdater, UnusedTensors, Optimizer, OwnedTape},
+    shapes::Shape,
+    tensor::{
+        AsArray, Cpu, HasErr, SampleTensor, SplitTape, Tensor, Tensor1D, Tensor2D, TensorFrom, Tensor0D,
+    },
+    tensor_ops::{MeanTo, TryDiv, Backward},
+};
+use std::f32;
+use std::time::Instant;
+
+const BATCH_SIZE: usize = 64;
+const STATE_SIZE: usize = 4;
+const INNER_SIZE: usize = 128;
+const ACTION_SIZE: usize = 2;
+
+/// Custom model struct
+#[derive(Clone)]
+struct Network<const IN: usize, const INNER: usize, const OUT: usize> {
+    l1: (Linear<IN, INNER, f32, Cpu>, ReLU),
+    mu: (Linear<INNER, OUT, f32, Cpu>, Tanh),
+    std: (Linear<INNER, OUT, f32, Cpu>, ReLU), // TODO: should this be SoftPlus?
+    value: Linear<INNER, OUT, f32, Cpu>,
+}
+
+// BuildModule lets you randomize a model's parameters
+impl<const IN: usize, const INNER: usize, const OUT: usize> nn::BuildModule<Cpu, f32>
+    for Network<IN, INNER, OUT>
+{
+    fn try_build(device: &Cpu) -> Result<Self, <Cpu as HasErr>::Err> {
+        Ok(Self {
+            l1: BuildModule::try_build(device)?,
+            mu: BuildModule::try_build(device)?,
+            std: BuildModule::try_build(device)?,
+            value: BuildModule::try_build(device)?,
+        })
+    }
+}
+
+// GradientUpdate lets you update a model's parameters using gradients
+impl<const IN: usize, const INNER: usize, const OUT: usize> GradientUpdate<Cpu, f32>
+    for Network<IN, INNER, OUT>
+{
+    fn update<U>(
+        &mut self,
+        updater: &mut U,
+        unused: &mut UnusedTensors,
+    ) -> Result<(), <Cpu as HasErr>::Err>
+    where
+        U: ParamUpdater<Cpu, f32>,
+    {
+        self.l1.update(updater, unused)?;
+        self.mu.update(updater, unused)?;
+        self.std.update(updater, unused)?;
+        self.value.update(updater, unused)?;
+        Ok(())
+    }
+}
+
+// impl Module for single item
+impl<const IN: usize, const INNER: usize, const OUT: usize, T: Tape<Cpu>> nn::Module<Tensor1D<IN, T>>
+    for Network<IN, INNER, OUT>
+{
+    type Output = (Tensor1D<OUT, T>, Tensor1D<OUT, T>, Tensor1D<OUT, T>);
+
+    fn forward(&self, x: Tensor1D<IN, T>) -> Self::Output {
+        let x = self.l1.forward(x);
+        (
+            self.mu.forward(x),
+            self.std.forward(x),
+            self.value.forward(x),
```

Comment:
How do clone `x`? `OwnedTape<Cpu>` does not implement `Copy` and I can not pass x by reference. I am not sure I understand the concept of Tape. Will try to google it another day...

Reply:
I believe that the way to do this is to do `x.retaped()` to clone the data of a tensor without cloning it's `Tape`. Since tapes serve to keep a record of what backward operations need to be done once `backward` is called, you will need to be sure that the output using the non-retaped x is always used to calculate the loss. Otherwise, the tape will not contain the operations done by `l1`, and `l1` will not be trained.

Reply:
I've actually decided that I like that tapes cannot be cloned from a design standpoint, because it prevents users from calculating gradients inefficiently, i.e. with multiple calls to backwards instead of only one. If you use `retaped` in Network's `forward`, you shouldn't encounter any issues, since you end up using all recombining all of it's outputs when calculating the gradient.